### ISSUE TITLE
clicking deltachat icon in swaybar's system tray offers inconsistent behavior (and can crash app)

### ISSUE THREAD

**Issue by dkg** [deltachat/deltachat-desktop#3989]
- OS: Debian stable (12.5) amd64
- Deltachat-desktop version: 1.46.1

I'm running `deltachat-desktop` under [sway](https://github.com/swaywm/sway), with no Xwayland or X11 support on the system at all.  I've launched deltachat by using electron's (chromium's) ozone backend for wayland:

```
deltachat-desktop --ozone-platform=wayland
```

When delta first starts up, it is *only* rendered as an icon in the system tray section of swaybar.  If i click that icon, the full deltachat interface appears.

If i click it again, the deltachat interface disappears.

But if i click it a third time, deltachat closes entirely (the app appears to crash), emitting the following message to stdout or stderr:

```
[3179:0628/212044.664566:FATAL:wayland_toplevel_window.cc(1123)] Should not be called with kMinimized state
Trace/breakpoint trap
```

When the app is running but minimized, right-clicking the icon in the system tray creates a menu with three items:

- Open DeltaChat
- Minimize
- Quit

if i choose "Open DeltaChat", that also causes the crash described above.  "Quit" appears to close the app cleanly.

Note that the menu doesn't interact well with sway -- it wants to consume the entire screen, and it doesn't share the screen at all well with the main app itself.

I'm happy to test or diagnose things if that would be useful.

**Comment by farooqkz** [comment]
As a workaround for this issue, use DeltaChat with Xwayland. I've been using it for sometime and it's fine.

**Comment by hpk42** [comment]
Well, dkg was describing a system setup that explicitely not uses wayland
so recommending to use wayland does not help much i am afraid.

**Comment by farooqkz** [comment]
@Simon-Laux There are other reports which specificly target wayland on my side. Any on your side?

**Comment by farooqkz** [comment]
> Well, dkg was describing a system setup that explicitely not uses wayland so recommending to use wayland does not help much i am afraid.
> […](#)
> On Mon, Jul 01, 2024 at 05:23 -0700, Farooq Karimi Zadeh wrote: As a workaround for this issue, use DeltaChat with Xwayland. I've been using it for sometime and it's fine. -- Reply to this email directly or view it on GitHub: [#3989 (comment)](https://github.com/deltachat/deltachat-desktop/issues/3989#issuecomment-2200004161) You are receiving this because you are subscribed to this thread. Message ID: ***@***.***>

The issue is with wayland. @dkg is already using sway which indicates he is using wayland as well. The problem's with native wayland support on Electron probably. What I suggested is a temporary workaround, using xwayland, till Electron is fixed or we fix it somehow.

**Comment by farooqkz** [comment]
BTW the user mentioning problems on wayland is @BehdaniSohrab. mentioning him with the hope to get more feedback on wayland related issues.

**Comment by behdanisohrab** [comment]
> BTW the user mentioning problems on wayland is @behdanisohrab. mentioning him with the hope to get more feedback on wayland related issues.

Most of the wayland issues on electron application is caused by chromium engine inside of it.

easiest way for enabling wayland support is by adding 
```
--enable-features=UseOzonePlatform,WaylandWindowDecorations --ozone-platform-hint=auto
```
to the desktop file or in front of appimage file.

i would provide more information for this problem later.

**Comment by dkg** [comment]
yes, i'm using wayland.  no, i'm not going to run XWayland on this system, for other reasons unrelated to DeltaChat.   as i said in the first message, i'm launching it with `--ozone-platform=wayland` specifically to use the wayland backend for electron.

With deltachat desktop 1.48.0, i've tested and the same error described here is still happening, along with the same message to stderr on the third click:

```
[1564:1115/082245.505552:FATAL:wayland_toplevel_window.cc(1113)] Should not be called with kMinimized state
```

**Comment by dkg** [comment]
btw, i've also tried running with `--enable-features=WaylandWindowDecorations` as recommended by @behdanisohrab but it seems to have no effect on the behavior i'm seeing.

**Comment by Simon-Laux** [comment]
we recently upgraded electron, is this still an issue in the current version (1.54.1)?

**Comment by farooqkz** [comment]
> > BTW the user mentioning problems on wayland is [@behdanisohrab](https://github.com/behdanisohrab). mentioning him with the hope to get more feedback on wayland related issues.
> 
> Most of the wayland issues on electron application is caused by chromium engine inside of it.
> 
> easiest way for enabling wayland support is by adding
> 
> ```
> --enable-features=UseOzonePlatform,WaylandWindowDecorations --ozone-platform-hint=auto
> ```
> 
> to the desktop file or in front of appimage file.
> 
> i would provide more information for this problem later.

Hello Sohrab. You said you would provide more information later.

**Comment by dkg** [comment]
this misbehavior seems to still be happening in deltachat-desktop 1.60.1 under wayland.

If i  click the icon in the system tray a third time, the app crashes with `Should not be called with kMinimized state`.

**Comment by link2xt** [comment]
I opened #5393 then realized it is a duplicate.

This is also reproducible with waybar instead of swaybar and with Niri or Hyprland instead of sway.

**Comment by akarelas** [comment]
I have KDE on Wayland, and Delta Chat (the flathub version) also always closes when I do this:

* launch Delta Chat (it sits in tray, no window appears)
* click tray icon (window appears)
* click tray icon (window disappears)
* click tray icon (application closes (according to ps), tray icon disappears, window doesn't open)

My distro is Void Linux.

This bug is the only reason I'm not using Delta Chat.

**Comment by akarelas** [comment]
> As a workaround for this issue, use DeltaChat with Xwayland. I've been using it for sometime and it's fine.

I'm running DeltaChat flatpak from flathub on a system that has Xwayland, and I'm still getting that bug. Disabling Wayland from flatpak's permission system (in order to force xwayland to take over) causes Delta Chat to not launch at all.

**Comment by akarelas** [comment]
> The issue is with wayland. [@dkg](https://github.com/dkg) is already using sway which indicates he is using wayland as well. The problem's with native wayland support on Electron probably. What I suggested is a temporary workaround, using xwayland, till Electron is fixed or we fix it somehow.

How can I use xwayland with the flathub version of Delta Chat?

**Comment by akarelas** [comment]
[Signal is also an Electron app](https://www.reddit.com/r/signal/comments/94suot/is_signal_an_electron_app/), yet doesn't have that problem. Why not copy what they did to your app?

**Comment by link2xt** [comment]
The error "Should not be called with kMinimized state" comes from here:
https://chromium.googlesource.com/chromium/src/+/refs/heads/main/ui/ozone/platform/wayland/host/wayland_toplevel_window.cc

I have commented out `mainWindow.window.minimize()` locally in packages/target-electron/src/tray.ts, this fixes the problem for me with `niri`. In niri I cannot minimize the window, so will need to run some proper desktop to test the fix for real.

Otherwise the crash happens in `showDeltaChat()`, when `mainWindow.window.show()` is called. Apparently chromium crashes if the window is minimized, hidden, then shown. Workaround is not to minimize it from the app, but maybe it will still crash if the user minimizes the app, then clicks the icon to hide minimized window, then clicks the icon to show it. We should probably unminimize the app if it is minimized and the icon is clicked.

**Comment by link2xt** [comment]
Made a PR with a fix: https://github.com/deltachat/deltachat-desktop/pull/5774
